File: Point.py
import math

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    # Operator overloading

    def __add__(self, other):
        x = self.x + other.x
        y = self.y + other.y
        return Point(x, y)

    def __truediv__(self, other):
        x = self.x/other
        y = self.y/other

        return Point(x, y)

    def __sub__(self, other):
        x = self.x - other.x
        y = self.y - other.y
        return Point(x, y)

    def __rmul__(self, other):
        x = self.x * other
        y = self.y * other
        return Point(x, y)

    def __mul__(self, other):
        x = self.x * other
        y = self.y * other
        return Point(x, y)

    def __str__(self):
        return "({0},{1})".format(self.x, self.y)

def distance_formula(p1, p2):
    """Returns distance between two points"""
    return (math.sqrt((p2.x - p1.x) ** 2 + (p2.y - p1.y) ** 2))

def magnitude(p):
    """Returns magnitude of a vector"""
    return math.sqrt(p.x**2 + p.y**2)

def dot_product(v1, v2):
    """Returns dot product of two vectors"""
    return v1.x * v2.x + v1.y * v2.y
def determinant(v1, v2):
    """Returns determinant of 2D matrix with two vectors"""
    return v1.x * v2.y - v2.x * v1.y

def slope(p1, p2):
    p = p2 - p1
    return (p.y/p.x)

def getPerpendicularVector(A, B, C):
    """Returns vector perpendicular to the angle bisector of the angle inscribed between neighboring points."""

    AB = B-A
    BC = C-B

    v1 = (AB)/(magnitude(AB))
    v2 = (BC)/(magnitude(BC))

    try:
        return ((v1+v2)/magnitude(v1+v2))
    except:
        return Point(1, 0)

File: frontend/public/Point.js
class Point {
    constructor(x, y) {
      this.x = x;
      this.y = y;
    }
    // Return a new Point that is the sum of this and another point.
    add(other) {
      return new Point(this.x + other.x, this.y + other.y);
    }
    // Return a new Point that is the difference of this and another point.
    subtract(other) {
      return new Point(this.x - other.x, this.y - other.y);
    }
    // Multiply by a scalar.
    multiply(scalar) {
      return new Point(this.x * scalar, this.y * scalar);
    }
    // Divide by a scalar.
    divide(scalar) {
      return new Point(this.x / scalar, this.y / scalar);
    }
    // Round the coordinates.
    round(ndigits = 0) {
      return new Point(Number(this.x.toFixed(ndigits)), Number(this.y.toFixed(ndigits)));
    }
    toString() {
      return `(${this.x}, ${this.y})`;
    }
  }
  
  function distance_formula(p1, p2) {
    return Math.sqrt(Math.pow(p2.x - p1.x, 2) + Math.pow(p2.y - p1.y, 2));
  }
  
  function magnitude(p) {
    return Math.sqrt(p.x * p.x + p.y * p.y);
  }
  
  function dot_product(v1, v2) {
    return v1.x * v2.x + v1.y * v2.y;
  }
  
  function determinant(v1, v2) {
    return v1.x * v2.y - v2.x * v1.y;
  }
  
  function slope(p1, p2) {
    return (p2.subtract(p1)).y / (p2.subtract(p1)).x;
  }
  
  /**
   * Returns a unit vector perpendicular to the angle bisector
   * of the angle formed by three points A, B, and C.
   */
  function getPerpendicularVector(A, B, C) {
    const AB = B.subtract(A);
    const BC = C.subtract(B);
    const v1 = AB.divide(magnitude(AB));
    const v2 = BC.divide(magnitude(BC));
    const sum = new Point(v1.x + v2.x, v1.y + v2.y);
    const magSum = magnitude(sum);
    if (magSum === 0) return new Point(1, 0);
    return sum.divide(magSum);
  }